BattleCity/sources/protocols/fieldmessage.cpp

```diff
@@ -0,0 +1,447 @@
+
+			if ((enumValue & setFieldsValue) == enumValue) {
+				Fields setField = static_cast<Fields>(enumValue);
+				if (FieldHasType(setField, INT_FIELD)) {
+					AddNewField(setField, intValue);
+				}
+				else {
+					throw logic_error("Field's type is not an int!");
+				}
+			}
+		}
+	}
+
+	void FieldMessage::AddNewField(Fields setField, int intValue) {
+		fields_[setField] = intValue;
+		get<int>(fields_[Fields::MESSAGE_SIZE]) += NEW_INT_FIELD_BYTES;
+	}
+
+	void FieldMessage::AddNewField(Fields setField, const string& stringValue) {
+		fields_[setField] = stringValue;
+		get<int>(fields_[Fields::MESSAGE_SIZE]) += stringValue.size();
+		get<int>(fields_[Fields::MESSAGE_SIZE]) += NEW_STR_FIELD_BYTES;
+	}
+
+	string FieldMessage::GetString(Fields field) const {
+		if (!FieldIsPresentInEnum(field)) throw out_of_range("No such field present in enum!");
+		if (!FieldHasType(field, STRING_FIELD)) throw logic_error("Field's type is not a string!");
+		if (!FieldIsSet(field)) throw out_of_range("Field is not set!");
+
+		return get<string>(fields_.at(field));
+	}
+
+	int FieldMessage::GetInt(Fields  field) const {
+		if (!FieldIsPresentInEnum(field)) throw out_of_range("No such field present in enum!");
+		if (!FieldHasType(field, INT_FIELD)) throw logic_error("Field's type is not an int!");
+		if (!FieldIsSet(field)) throw out_of_range("Field is not set!");
+
+		return get<int>(fields_.at(field));
+	}
+
+	bool FieldMessage::Has(Fields fields) const {
+		FieldsUnderlyingType hasFieldsValue = static_cast<FieldsUnderlyingType>(fields);
+
+		for (size_t i = 0; i < enums_masks_.size(); ++i) {
+			FieldsUnderlyingType enumValue = enums_masks_[i];
+
+			if ((enumValue & hasFieldsValue) == enumValue) {
+				Fields hasField = static_cast<Fields>(enumValue);
+				if (!FieldIsSet(hasField)) {
+					return false;
+				}
+			}
+		}
+
+		return true;
+	}
+
+	void FieldMessage::DeleteField(Fields field) {
+		if (FieldIsPresentInEnum(field)) {
+			if (field == Fields::MESSAGE_SIZE || field == Fields::CORR_ID) throw logic_error("Cannot delete this field!");
+
+			auto deleteIterator = fields_.find(field);
+			if (deleteIterator != fields_.end()) {
+				RemoveField(deleteIterator, field);
+			}
+		}
+	}
+
+	void FieldMessage::RemoveField(map<Fields, SupportedVariant>::iterator deleteIterator, Fields field) {
+		
+		int bytesRemoved = 0;
+		auto& supportedVariant = (*deleteIterator).second;
+		if (holds_alternative<string>(supportedVariant)) {
+			bytesRemoved += NEW_STR_FIELD_BYTES;
+			bytesRemoved += get<string>(supportedVariant).size();
+		} else
+		if (holds_alternative<int>(supportedVariant)) {
+			bytesRemoved += NEW_INT_FIELD_BYTES;
+		} else {
+			throw logic_error("Unknown variant removal!");
+		}
+
+		fields_.erase(deleteIterator);
+		get<int>(fields_[Fields::MESSAGE_SIZE]) -= bytesRemoved;
+	}
+
+	void FieldMessage::WriteMessageSize(stringstream& ss, size_t msgSize) const {
+		using cchar = const char;
+		unsigned long long promoted = msgSize;
```

Comment:
`std::uint64_t` is 8-bit unsigned integer 
```suggestion
		std::uint64_t promoted = msgSize;
```